### DISCUSSION TITLE
Python API parity for COPY TO operations

### DISCUSSION THREAD

**Discussion by tconbeer** [duckdb/duckdb#8896] [Ideas]
I maintain [Harlequin](https://harlequin.sh). I have an Export UI that allows users to export data from their queries to CSV, Parquet, or JSON files.

I noticed that the Python API lacks features that are present in the SQL API for `COPY`. Namely:

1. There is no Python API for `write_json` (`COPY my_table to 'output.json'`)
2. `write_csv` lacks the following global options for `copy ... to`:
    - `partition_by`
    - `allow_overwrite`
    - `per_thread_output`
    - `use_tmp_file`

3. In addition, `write_parquet` lacks the following parquet-specific options for `copy .. to`:
    - `row_group_size` or `row_group_size_bytes`
    - `field_ids`

Since my app is written in Python, it would be nice to have a Python API for these features, instead of having to compose a SQL string.

If it's of interest, I'd be happy to contribute this feature.

**Comment by Mause** [comment]
Hey there Ted, if you want to have a go at adding features you're very welcome to - feel free to hit us up on discord if you have issues getting started

**Comment by iangow** [comment]
Ditto for `COPY ... FROM`. One can do `duckdb.from_csv_auto(StringIO(p.read().decode(encoding))` where `p` is a stream from a remote source, but RAM usage is huge. One can't easily make `p` outside Python, so it's very valuable to support this within the API. Something like `duckdb.copy_from()` would help. (With PostgreSQL, one has `copy_expert()` to do this.)

**Comment by jzavala-gonzalez** [comment]
Hey all, I'm currently attempting to add the CSV parameters that Ted listed and have some progress over [at a branch on my fork](https://github.com/jzavala-gonzalez/duckdb/tree/python-write-csv-options). Haven't opened a PR yet because I'm still figuring out how to compile and test it. If anyone is more comfortable doing so, please feel free to take a look! The piece I'm more uncertain about is how it handles the 'partition_by' argument, and checking if/how the stubs need to be regenerated.

**Comment by jzavala-gonzalez** [comment]
Opened a PR at https://github.com/duckdb/duckdb/pull/10382 . CI still needs to do its thing but testing locally the parameters work okay. An example script I used:

```python
import duckdb

con = duckdb.connect(':memory:')

con.execute(
    '''
create or replace table penguins as (
select * from 'penguins.csv'
)
'''
)

res = con.sql('select * from penguins')
print(res)

res.write_csv(
    'penguins_out.csv',
    partition_by=['species', 'sex'], overwrite=True,
    #   per_thread_output=True, # can't enable at same time as partition_by
)

con.close()

```

**Comment by jzavala-gonzalez** [comment]
The PR for the write_csv parameters just got merged! I'll let you know if I find some time to work on the other two functions
